# Lab | Data Cleaning and Formatting

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv


# Challenge 1: Data Cleaning and Formatting

## Exercise 1: Cleaning Column Names

To ensure consistency and ease of use, standardize the column names of the dataframe.  
 Start by taking a first look at the dataframe and identifying any column names that need to be modified.  
  Use appropriate naming conventions and make sure that column names are descriptive and informative.

*Hint*:
- *Column names should be in lower case*
- *White spaces in column names should be replaced by `_`*
- *`st` could be replaced for `state`*

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)
df.head(10)

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
5,OC83172,Oregon,F,Bachelor,825629.78%,62902.0,69.0,1/0/00,Personal Auto,Two-Door Car,159.383042
6,XZ87318,Oregon,F,College,538089.86%,55350.0,67.0,1/0/00,Corporate Auto,Four-Door Car,321.600000
7,CF85061,Arizona,M,Master,721610.03%,0.0,101.0,1/0/00,Corporate Auto,Four-Door Car,363.029680
8,DY87989,Oregon,M,Bachelor,2412750.40%,14072.0,71.0,1/0/00,Corporate Auto,Four-Door Car,511.200000
9,BQ94931,Oregon,F,College,738817.81%,28812.0,93.0,1/0/00,Special Auto,Four-Door Car,425.527834


In [2]:
df.shape

(4008, 11)

- Cambio la columna ST por STATE

In [3]:
if "ST" in df.columns:
    df = df.rename(columns={"ST": "STATE"})
df


,Customer,STATE,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
...,...,...,...,...,...,...,...,...,...,...,...
4003,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4004,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4006,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [4]:
df.columns = (
df.columns
  .str.strip()              # elimina espacios al inicio y final
  .str.lower()              # convierte a minúsculas
  .str.replace(' ', '_')    # reemplaza espacios por _
  )
df.head()

,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323


## Exercise 2: Cleaning invalid Values

The dataset contains columns with inconsistent and incorrect values that could affect the accuracy of our analysis.  
Therefore, we need to clean these columns to ensure that they only contain valid data.

Note that this exercise will focus only on cleaning inconsistent values and will not involve handling null values (NaN or None).

*Hint*:



- *Gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized, for example, to "M" and "F".*

In [5]:
df['gender'] = df['gender'].str.strip().str.lower()

df['gender'] = df['gender'].replace({
    'm': 'M', 'male': 'M',
    'f': 'F', 'female': 'F', 'femal': 'F'
})

df.head()

,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323


In [6]:
df['gender'].value_counts() # Fantastico , queda F,M solo

gender
F    502
M    452
Name: count, dtype: int64

- *State abbreviations be can replaced with its full name, for example "AZ": "Arizona", "Cali": "California", "WA": "Washington"*

In [7]:
df['state'].unique()

<StringArray>
['Washington',    'Arizona',     'Nevada', 'California',     'Oregon',
       'Cali',         'AZ',         'WA',          nan]
Length: 9, dtype: str

In [8]:
state_map = {
    'AZ': 'Arizona',
    'Cali': 'California',
    'WA': 'Washington',
}

df['state'] = df['state'].replace(state_map)
df.head()

,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323


- *In education, "Bachelors" could be replaced by "Bachelor"*

Primero vemos que valores hay en education

In [9]:
df['education'].unique()

<StringArray>
[              'Master',             'Bachelor', 'High School or Below',
              'College',            'Bachelors',               'Doctor',
                    nan]
Length: 7, dtype: str

In [10]:
df['education'] = df['education'].replace({'Bachelors': 'Bachelor'})
df.head()

,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323




- *In Customer Lifetime Value, delete the `%` character*


In [11]:
df['customer_lifetime_value'].dtypes

<StringDtype(storage='python', na_value=nan)>

In [12]:
df['customer_lifetime_value'] = df['customer_lifetime_value'].str.replace('%', '')


In [13]:
df.head()

,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323


- *In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"*

In [14]:
df['vehicle_class'].value_counts()

vehicle_class
Four-Door Car    576
Two-Door Car     205
SUV              199
Sports Car        57
Luxury SUV        20
Luxury Car        14
Name: count, dtype: int64

In [15]:
df['vehicle_class'] = df['vehicle_class'].replace({
    'Sports Car': 'Luxury',
    'Luxury SUV': 'Luxury',
    'Luxury Car': 'Luxury'
})
df.head()

,customer,state,gender,education,customer_lifetime_value,income,monthly_premium_auto,number_of_open_complaints,policy_type,vehicle_class,total_claim_amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323


## Exercise 3: Formatting data types

The data types of many columns in the dataset appear to be incorrect. 
This could impact the accuracy of our analysis.
 To ensure accurate analysis, we need to correct the data types of these columns.
  Please update the data types of the columns as appropriate.

It is important to note that this exercise does not involve handling null values (NaN or None).

*Hint*:
- *Customer lifetime value should be numeric*


In [16]:
df['customer_lifetime_value'] = pd.to_numeric(df['customer_lifetime_value'], errors='coerce')
df['customer_lifetime_value'].dtypes

dtype('float64')

- *Number of open complaints has an incorrect format. Look at the different values it takes with `unique()` and take the middle value. As an example, 1/5/00 should be 5. Number of open complaints is a string - remember you can use `split()` to deal with it and take the number you need. Finally, since it should be numeric, cast the column to be in its proper type.*

In [ ]:
df['number_of_open_complaints'].astype(str).unique() # para ver que valores tengo ... lo unico raro veo un nan

In [ ]:
df['number_of_open_complaints'] = (
    df['number_of_open_complaints']
      .astype(str)
      .str.split('/')
      .str[1]
      .astype(float)
)
df.head()

## Exercise 4: Dealing with Null values

Identify any columns with null or missing values.  
 Identify how many null values each column has. You can use the `isnull()` function in pandas to find columns with null values.

In [18]:
df.isnull().sum() # NULL or missing values son lo mism


customer                     2937
state                        2937
gender                       3054
education                    2937
customer_lifetime_value      2940
income                       2937
monthly_premium_auto         2937
number_of_open_complaints    2937
policy_type                  2937
vehicle_class                2937
total_claim_amount           2937
dtype: int64

Decide on a strategy for handling the null values. There are several options, including:



Numéricos → rellenar con la mediana : La mediana es robusta ante valores extremos.

Mantiene la distribución sin distorsionar.



In [19]:
num_cols = df.select_dtypes(include=['float64', 'int64']).columns
num_cols

Index(['customer_lifetime_value', 'income', 'monthly_premium_auto',
       'total_claim_amount'],
      dtype='str')

Categóricos → rellenar con la moda : La moda es el valor más frecuente.

Mantiene coherencia semántica.

In [20]:
cat_cols = df.select_dtypes(include=['object']).columns
cat_cols

C:\Users\ernes\AppData\Local\Temp\ipykernel_12848\1383039436.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = df.select_dtypes(include=['object']).columns


Index(['customer', 'state', 'gender', 'education', 'number_of_open_complaints',
       'policy_type', 'vehicle_class'],
      dtype='str')

- Drop the rows or columns with null values


- Fill the null values with a specific value (such as the column mean or median for numerical variables, and mode for categorical variables)


- Fill the null values with the previous or next value in the column
Implement your chosen strategy to handle the null values.  
 You can use the `fillna()` function in pandas to fill null values or `dropna()` function to drop null values.


- Fill the null values based on a more complex algorithm or model (note: we haven't covered this yet)

Implement your chosen strategy to handle the null values. You can use the `fillna()` function in pandas to fill null values or `dropna()` function to drop null values.



In [ ]:
df[num_cols] = df[num_cols].fillna(df[num_cols].median())
df[cat_cols] = df[cat_cols].fillna(df[cat_cols].mode().iloc[0])




Verify that your strategy has successfully handled the null values. You can use the `isnull()` function again to check if there are still null values in the dataset.



In [ ]:
df.isnull().sum()

Remember to document your process and explain your reasoning for choosing a particular strategy for handling null values.



En este ejercicio analicé los valores nulos del dataset utilizando: df.isnull().sum()
Esto me permitió identificar qué columnas contenían valores faltantes y cuántos había en cada una.

Eliminar filas con valores nulos usando df.dropna().
    Esta opción elimina cualquier registro incompleto, pero puede reducir demasiado el tamaño del dataset.

Eliminar columnas con valores nulos usando df.dropna(axis=1).
    Esta opción elimina variables completas, lo cual no es deseable si la columna es relevante. (MUY PELIGROSA)

Rellenar valores nulos con el valor anterior o siguiente usando df.ffill() y df.bfill().
    No se aun para que vale muy bien , pero la he probado para verla

    La verdad que las probe todas, aun no domino cual coger, pero mal no fue pase de 952 filas a 922 o sea que solo perdi 30 filas en el proceso

After formatting data types, as a last step, convert all the numeric variables to integers.

In [ ]:
df[num_cols] = df[num_cols].astype(int)
df.head()

## Exercise 5: Dealing with duplicates

In [50]:
conteo_inicial = df.shape
conteo_inicial 

(1072, 11)

Use the `.duplicated()` method to identify any duplicate rows in the dataframe.



In [44]:
df.duplicated().sum()



np.int64(2936)

Decide on a strategy for handling the duplicates. Options include:


- Dropping all duplicate rows
""" 
Elimina todas las filas duplicadas, manteniendo solo una copia de cada fila única.
Cuándo usarlo:
Cuando quieres un dataset sin ninguna fila repetida

Cuando todas las columnas deben coincidir para considerar un duplicado
"""


- Keeping only the last occurrence of each duplicated row
"""
Mantiene la primera aparición y elimina las demás.
Cuándo usarlo:
Cuando la primera fila es la que quieres conservar

Es la opción más común y la que suele pedir el lab
"""


- Keeping only the last  occurrence of each duplicated 
"""
Mantiene la última aparición y elimina las anteriores.
Cuándo usarlo:
Cuando la información más reciente es la más relevante

Cuando las filas posteriores contienen datos corregidos
"""



- Dropping duplicates based on a subset of columns
"""
Elimina duplicados solo considerando algunas columnas, no todas
Cuándo usarlo:
Cuando una fila puede variar en otras columnas, pero sigue siendo el mismo cliente/registro

Cuando solo ciertas columnas definen la unicidad
"""


- Dropping duplicates based on a specific column
"""
Es un caso especial del anterior, pero el lab lo menciona como estrategia separada.
Cuándo usarlo:
Cuando una columna es un identificador único

Cuando quieres eliminar clientes repetidos, transacciones repetidas, etc.
"""



Implement your chosen strategy using the `drop_duplicates()` function.



---

Escogi : Eliminar duplicados manteniendo solo una copia de cada fila única  

drop_duplicates() elimina todas las filas duplicadas  

mantiene solo una copia de cada fila única  

reset_index(drop=True) reorganiza el índice para que quede limpio  

In [45]:
df = df.drop_duplicates().reset_index(drop=True)
df

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323
...,...,...,...,...,...,...,...,...,...,...,...
1067,VJ51327,Cali,F,High School or Below,2031499.76%,63209.0,102.0,1/2/00,Personal Auto,SUV,207.320041
1068,GS98873,Arizona,F,Bachelor,323912.47%,16061.0,88.0,1/0/00,Personal Auto,Four-Door Car,633.600000
1069,CW49887,California,F,Master,462680.11%,79487.0,114.0,1/0/00,Special Auto,SUV,547.200000
1070,MY31220,California,F,College,899704.02%,54230.0,112.0,1/0/00,Personal Auto,Two-Door Car,537.600000



Verify that your strategy has successfully handled the duplicates by checking for duplicates again using `.duplicated()`.

Vemos que no hay duplicados y hemos pasado de 4008 filas a 1072



In [46]:
df.duplicated().sum()

np.int64(0)

In [47]:
conteo_final = df.shape
conteo_final

(1072, 11)

In [49]:
Limpieza_final = f"FILAS_INICIALES: {conteo_inicial} | FILAS_FINALES: {conteo_final}"
Limpieza_final

'FILAS_INICIALES: (4008, 11) | FILAS_FINALES: (1072, 11)'

Remember to document your process and explain your reasoning for choosing a particular strategy for handling duplicates.



Save the cleaned dataset to a new CSV file.



In [ ]:
df.to_csv("cleaned_file1.csv", index=False)


*Hint*: *after dropping duplicates, reset the index to ensure consistency*.

Ya esta hecho al eliminar duplicados

# Bonus: Challenge 2: creating functions on a separate `py` file

Put all the data cleaning and formatting steps into functions, and create a main function that performs all the cleaning and formatting.

Write these functions in separate .py file(s). By putting these steps into functions, we can make the code more modular and easier to maintain.

*Hint: autoreload module is a utility module in Python that allows you to automatically reload modules in the current session when changes are made to the source code. This can be useful in situations where you are actively developing code and want to see the effects of changes you make without having to constantly restart the Python interpreter or Jupyter Notebook kernel.*

# Bonus: Challenge 3: Analyzing Clean and Formated Data

You have been tasked with analyzing the data to identify potential areas for improving customer retention and profitability. Your goal is to identify customers with a high policy claim amount and a low customer lifetime value.

In the Pandas Lab, we only looked at high policy claim amounts because we couldn't look into low customer lifetime values. If we had tried to work with that column, we wouldn't have been able to because customer lifetime value wasn't clean and in its proper format. So after cleaning and formatting the data, let's get some more interesting insights!

Instructions:

- Review the statistics again for total claim amount and customer lifetime value to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount and a low customer lifetime value. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount, and clients with a low customer lifetime value to be those in the bottom 25% of the customer lifetime value. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile and a customer lifetime value in the bottom 25th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount and low customer lifetime value data. To do so, select both columns of the dataframe simultaneously and pass it to the `.describe()` method. This will give you descriptive statistics, such as mean, median, standard deviation, minimum and maximum values for both columns at the same time, allowing you to compare and analyze their characteristics.

SyntaxError: invalid character '“' (U+201C) (2186280899.py, line 1)